### ISSUE TITLE
Cross compiling error with cmake

### ISSUE THREAD

**Issue by apmanol** [google/benchmark#468]
Hello, this issue is probable related to #436 , during cross compiling I have the following problem:
```
cmake -DCMAKE_TOOLCHAIN_FILE=/home/manap/Projects/xpu/nmsngv3/tools/share/cmake/toolchain-mips-wibasip.cmake -DBENCHMARK_ENABLE_TESTING=OFF .. 
-- The C compiler identification is GNU 4.8.1
-- The CXX compiler identification is GNU 4.8.1
-- Check for working C compiler: /opt/cross-mips-wibasip/cross.mips/bin/mips-wibasip-linux-gnu-gcc
-- Check for working C compiler: /opt/cross-mips-wibasip/cross.mips/bin/mips-wibasip-linux-gnu-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /opt/cross-mips-wibasip/cross.mips/bin/mips-wibasip-linux-gnu-g++
-- Check for working CXX compiler: /opt/cross-mips-wibasip/cross.mips/bin/mips-wibasip-linux-gnu-g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
crosssss compiling
-- Found Git: /usr/bin/git (found version "2.9.5") 
-- git Version: v1.2.0-491360b8-dirty
-- Version: 1.2.0
-- Performing Test HAVE_CXX_FLAG_STD_CXX11
-- Performing Test HAVE_CXX_FLAG_STD_CXX11 - Success
-- Performing Test HAVE_CXX_FLAG_WALL
-- Performing Test HAVE_CXX_FLAG_WALL - Success
-- Performing Test HAVE_CXX_FLAG_WEXTRA
-- Performing Test HAVE_CXX_FLAG_WEXTRA - Success
-- Performing Test HAVE_CXX_FLAG_WSHADOW
-- Performing Test HAVE_CXX_FLAG_WSHADOW - Success
-- Performing Test HAVE_CXX_FLAG_WERROR
-- Performing Test HAVE_CXX_FLAG_WERROR - Success
-- Performing Test HAVE_CXX_FLAG_PEDANTIC
-- Performing Test HAVE_CXX_FLAG_PEDANTIC - Success
-- Performing Test HAVE_CXX_FLAG_PEDANTIC_ERRORS
-- Performing Test HAVE_CXX_FLAG_PEDANTIC_ERRORS - Success
-- Performing Test HAVE_CXX_FLAG_WSHORTEN_64_TO_32
-- Performing Test HAVE_CXX_FLAG_WSHORTEN_64_TO_32 - Failed
-- Performing Test HAVE_CXX_FLAG_WFLOAT_EQUAL
-- Performing Test HAVE_CXX_FLAG_WFLOAT_EQUAL - Success
-- Performing Test HAVE_CXX_FLAG_FSTRICT_ALIASING
-- Performing Test HAVE_CXX_FLAG_FSTRICT_ALIASING - Success
-- Performing Test HAVE_CXX_FLAG_WZERO_AS_NULL_POINTER_CONSTANT
-- Performing Test HAVE_CXX_FLAG_WZERO_AS_NULL_POINTER_CONSTANT - Success
-- Performing Test HAVE_CXX_FLAG_WSTRICT_ALIASING
-- Performing Test HAVE_CXX_FLAG_WSTRICT_ALIASING - Success
-- Performing Test HAVE_CXX_FLAG_WD654
-- Performing Test HAVE_CXX_FLAG_WD654 - Failed
-- Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY
-- Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY - Failed
-- Performing Test HAVE_CXX_FLAG_COVERAGE
-- Performing Test HAVE_CXX_FLAG_COVERAGE - Success
-- Performing Test HAVE_STD_REGEX
CMake Warning at cmake/CXXFeatureCheck.cmake:37 (message):
  If you see build failures due to cross compilation, try setting
  HAVE_STD_REGEX to 0
Call Stack (most recent call first):
  CMakeLists.txt:180 (cxx_feature_check)


-- Performing Test HAVE_STD_REGEX -- success
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Performing Test HAVE_POSIX_REGEX
CMake Warning at cmake/CXXFeatureCheck.cmake:37 (message):
  If you see build failures due to cross compilation, try setting
  HAVE_POSIX_REGEX to 0
Call Stack (most recent call first):
  CMakeLists.txt:182 (cxx_feature_check)


-- Performing Test HAVE_POSIX_REGEX -- success
-- Performing Test HAVE_STEADY_CLOCK
CMake Warning at cmake/CXXFeatureCheck.cmake:37 (message):
  If you see build failures due to cross compilation, try setting
  HAVE_STEADY_CLOCK to 0
Call Stack (most recent call first):
  CMakeLists.txt:190 (cxx_feature_check)


-- Performing Test HAVE_STEADY_CLOCK -- success
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Configuring done
-- Generating done
-- Build files have been written to: /projects/Thirdparty/benchmark/builds_mips2
/benchmark/builds_mips2 (master *%)$ make VERBOSE=1 
/usr/bin/cmake -H/projects/Thirdparty/benchmark -B/projects/Thirdparty/benchmark/builds_mips2 --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /projects/Thirdparty/benchmark/builds_mips2/CMakeFiles /projects/Thirdparty/benchmark/builds_mips2/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/projects/Thirdparty/benchmark/builds_mips2'
make -f src/CMakeFiles/benchmark.dir/build.make src/CMakeFiles/benchmark.dir/depend
make[2]: Entering directory '/projects/Thirdparty/benchmark/builds_mips2'
cd /projects/Thirdparty/benchmark/builds_mips2 && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /projects/Thirdparty/benchmark /projects/Thirdparty/benchmark/src /projects/Thirdparty/benchmark/builds_mips2 /projects/Thirdparty/benchmark/builds_mips2/src /projects/Thirdparty/benchmark/builds_mips2/src/CMakeFiles/benchmark.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/projects/Thirdparty/benchmark/builds_mips2'
make -f src/CMakeFiles/benchmark.dir/build.make src/CMakeFiles/benchmark.dir/build
make[2]: Entering directory '/projects/Thirdparty/benchmark/builds_mips2'
[  6%] Building CXX object src/CMakeFiles/benchmark.dir/benchmark.cc.o
cd /projects/Thirdparty/benchmark/builds_mips2/src && /opt/cross-mips-wibasip/cross.mips/bin/mips-wibasip-linux-gnu-g++  -DHAVE_POSIX_REGEX -DHAVE_STD_REGEX -DHAVE_STEADY_CLOCK -I/projects/Thirdparty/benchmark/include -I/projects/Thirdparty/benchmark/src -I/projects/Thirdparty/benchmark/src/../include  -std=c++11  -Wall  -Wextra  -Wshadow  -pedantic  -pedantic-errors  -Wfloat-equal  -fstrict-aliasing  -Wzero-as-null-pointer-constant  -Wstrict-aliasing   -o CMakeFiles/benchmark.dir/benchmark.cc.o -c /projects/Thirdparty/benchmark/src/benchmark.cc
^Csrc/CMakeFiles/benchmark.dir/build.make:62: recipe for target 'src/CMakeFiles/benchmark.dir/benchmark.cc.o' failed

```

And the question is should  flags `-DHAVE_POSIX_REGEX -DHAVE_STD_REGEX`  should be enabled since it failed? And the second one, is how can choose which one from cmake option. The attempts with `-DRUN_HAVE_POSIX_REGEX=0`  failed.

**Comment by dominichamon** [comment]
did you try what the warning says?

"CMake Warning at cmake/CXXFeatureCheck.cmake:37 (message):

  If you see build failures due to cross compilation, try setting
  HAVE_STD_REGEX to 0

"


Dominic Hamon | Google
*There are no bad ideas; only good ideas that go horribly wrong.*

**Comment by apmanol** [comment]
Indeed that was my second attempt:
```
$cmake -DCMAKE_TOOLCHAIN_FILE=/home/manap/Projects/xpu/nmsngv3/tools/share/cmake/toolchain-mips-wibasip.cmake -DBENCHMARK_ENABLE_TESTING=OFF -DHAVE_POSIX_REGEX=0  .. 
...
-- Performing Test HAVE_STD_REGEX
CMake Warning at cmake/CXXFeatureCheck.cmake:37 (message):
  If you see build failures due to cross compilation, try setting
  HAVE_STD_REGEX to 0
Call Stack (most recent call first):
  CMakeLists.txt:180 (cxx_feature_check)


-- Performing Test HAVE_STD_REGEX -- success
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Performing Test HAVE_STEADY_CLOCK
CMake Warning at cmake/CXXFeatureCheck.cmake:37 (message):
  If you see build failures due to cross compilation, try setting
  HAVE_STEADY_CLOCK to 0
Call Stack (most recent call first):
  CMakeLists.txt:190 (cxx_feature_check)


-- Performing Test HAVE_STEADY_CLOCK -- success
[  6%] Building CXX object src/CMakeFiles/benchmark.dir/benchmark.cc.o
cd /projects/Thirdparty/benchmark/builds_mips2/src && /opt/cross-mips-wibasip/cross.mips/bin/mips-wibasip-linux-gnu-g++  -DHAVE_POSIX_REGEX -DHAVE_STD_REGEX -DHAVE_STEADY_CLOCK -I/projects/Thirdparty/benchmark/include -I/projects/Thirdparty/benchmark/src -I/projects/Thirdparty/benchmark/src/../include  -std=c++11  -Wall  -Wextra  -Wshadow  -pedantic  -pedantic-errors  -Wfloat-equal  -fstrict-aliasing  -Wzero-as-null-pointer-constant  -Wstrict-aliasing   -o CMakeFiles/benchmark.dir/benchmark.cc.o -c /projects/Thirdparty/benchmark/src/benchmark.cc

```

**Comment by Maratyszcza** [comment]
Same error when cross-compiling for Android (using `android.toolchain.cmake` from Android NDK).

**Comment by EricWF** [comment]
Are you setting `-DCMAKE_SYSTEM_NAME=foo`?

**Comment by apmanol** [comment]
No, I am not aware of this option. What `foo` should be?

**Comment by Maratyszcza** [comment]
@EricWF `android.toolchain.cmake` from Android NDK sets this variable (and `CMAKE_SYSTEM_PROCESSOR` too) automatically.

**Comment by lordalcol** [comment]
I am currently trying to build for Android using:

`
cmake ../ -G Ninja -DCMAKE_SYSTEM_NAME=Android -DANDROID_ABI=arm64-v8a -DCMAKE_TOOLCHAIN_FILE=%ANDROID_NDK%/build/cmake/android.toolchain.cmake -DANDROID_TOOLCHAIN=clang -DCMAKE_BUILD_TYPE=Release -DBENCHMARK_ENABLE_TESTING=false -DANDROID_STL=gnustl_static -DANDROID_NATIVE_API_LEVEL=android-22 -DBENCHMARK_USE_LIBCXX=ON -DBENCHMARK_ENABLE_EXCEPTIONS=Off
ninja
`

But I get the error:
`
../src/sysinfo.cc:213:23: error: no member named 'stoul' in namespace 'std'
    CPUMask Mask(std::stoul(Part, nullptr, 16));
                 ~~~~~^
../src/sysinfo.cc:219:11: error: invalid operands to binary expression ('int' and 'void')
    total += CountBits(Val.substr(0, Pos));
    ~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/sysinfo.cc:223:11: error: invalid operands to binary expression ('int' and 'void')
    total += CountBits(Val);
    ~~~~~ ^  ~~~~~~~~~~~~~~
../src/sysinfo.cc:387:26: error: no member named 'stoi' in namespace 'std'
        int CurID = std::stoi(value);
                    ~~~~~^
../src/sysinfo.cc:460:41: error: no member named 'stod' in namespace 'std'
        double cycles_per_second = std::stod(value) * 1000000.0;
                                   ~~~~~^
../src/sysinfo.cc:465:27: error: no member named 'stod' in namespace 'std'
        bogo_clock = std::stod(value) * 1000000.0;
                     ~~~~~^
6 errors generated.
`

If I use c++_static instead it compiles but when I try to link it against my native lib it throws hundreds of errors like: `Error:(24) undefined reference to `std::__ndk1::clog'` or `Error:(538) undefined reference to `std::__ndk1::cerr'`

**Comment by mkwork** [comment]
> I am currently trying to build for Android using:
> 
> ```
> cmake ../ -G Ninja -DCMAKE_SYSTEM_NAME=Android -DANDROID_ABI=arm64-v8a -DCMAKE_TOOLCHAIN_FILE=%ANDROID_NDK%/build/cmake/android.toolchain.cmake -DANDROID_TOOLCHAIN=clang -DCMAKE_BUILD_TYPE=Release -DBENCHMARK_ENABLE_TESTING=false -DANDROID_STL=gnustl_static -DANDROID_NATIVE_API_LEVEL=android-22 -DBENCHMARK_USE_LIBCXX=ON -DBENCHMARK_ENABLE_EXCEPTIONS=Off
> ninja
> ```
> 
> But I get the error:
> 
> ```
> ../src/sysinfo.cc:213:23: error: no member named 'stoul' in namespace 'std'
>     CPUMask Mask(std::stoul(Part, nullptr, 16));
>                  ~~~~~^
> ../src/sysinfo.cc:219:11: error: invalid operands to binary expression ('int' and 'void')
>     total += CountBits(Val.substr(0, Pos));
>     ~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ../src/sysinfo.cc:223:11: error: invalid operands to binary expression ('int' and 'void')
>     total += CountBits(Val);
>     ~~~~~ ^  ~~~~~~~~~~~~~~
> ../src/sysinfo.cc:387:26: error: no member named 'stoi' in namespace 'std'
>         int CurID = std::stoi(value);
>                     ~~~~~^
> ../src/sysinfo.cc:460:41: error: no member named 'stod' in namespace 'std'
>         double cycles_per_second = std::stod(value) * 1000000.0;
>                                    ~~~~~^
> ../src/sysinfo.cc:465:27: error: no member named 'stod' in namespace 'std'
>         bogo_clock = std::stod(value) * 1000000.0;
>                      ~~~~~^
> 6 errors generated.
> ```
> 
> If I use c++_static instead it compiles but when I try to link it against my native lib it throws hundreds of errors like: `Error:(24) undefined reference to `std::__ndk1::clog'`or`Error:(538) undefined reference to `std::__ndk1::cerr'`

It's not seems to be related to this concrete issue. Also it looks like this issue is  partially fixed in `1.4.1`. At least while `CMAKE_CROSSCOMPILING=ON` manually defined.

**Comment by apmanol** [comment]
Currently with commit:6a5c379  and 

```
cmake -H. -Bbuild_perf_mips \
-DCMAKE_BUILD_TYPE=RelWithDebInfo \
-DCMAKE_TOOLCHAIN_FILE=../tools/share/cmake/toolchain-mips.cmake \
-DCMAKE_CXX_STANDARD=11 \
-DHAVE_POSIX_REGEX=1 \
-DHAVE_STEADY_CLOCK=1
```
works!

```
 Performing Test HAVE_CXX_FLAG_STD_CXX11
-- Performing Test HAVE_CXX_FLAG_STD_CXX11 - Success
-- Performing Test HAVE_CXX_FLAG_WALL
-- Performing Test HAVE_CXX_FLAG_WALL - Success
-- Performing Test HAVE_CXX_FLAG_WEXTRA
-- Performing Test HAVE_CXX_FLAG_WEXTRA - Success
-- Performing Test HAVE_CXX_FLAG_WSHADOW
-- Performing Test HAVE_CXX_FLAG_WSHADOW - Success
-- Performing Test HAVE_CXX_FLAG_WERROR
-- Performing Test HAVE_CXX_FLAG_WERROR - Success
-- Performing Test HAVE_CXX_FLAG_PEDANTIC
-- Performing Test HAVE_CXX_FLAG_PEDANTIC - Success
-- Performing Test HAVE_CXX_FLAG_PEDANTIC_ERRORS
-- Performing Test HAVE_CXX_FLAG_PEDANTIC_ERRORS - Success
-- Performing Test HAVE_CXX_FLAG_WSHORTEN_64_TO_32
-- Performing Test HAVE_CXX_FLAG_WSHORTEN_64_TO_32 - Failed
-- Performing Test HAVE_CXX_FLAG_FSTRICT_ALIASING
-- Performing Test HAVE_CXX_FLAG_FSTRICT_ALIASING - Success
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED_DECLARATIONS
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED_DECLARATIONS - Success
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED - Success
-- Performing Test HAVE_CXX_FLAG_WSTRICT_ALIASING
-- Performing Test HAVE_CXX_FLAG_WSTRICT_ALIASING - Success
-- Performing Test HAVE_CXX_FLAG_WD654
-- Performing Test HAVE_CXX_FLAG_WD654 - Failed
-- Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY
-- Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY - Failed
-- Performing Test HAVE_CXX_FLAG_COVERAGE
-- Performing Test HAVE_CXX_FLAG_COVERAGE - Success
-- Performing Test HAVE_STD_REGEX
CMake Warning at build_perf_mips/_deps/benchmark-src/cmake/CXXFeatureCheck.cmake:38 (message):
  If you see build failures due to cross compilation, try setting
  HAVE_STD_REGEX to 0
Call Stack (most recent call first):
  build_perf_mips/_deps/benchmark-src/CMakeLists.txt:246 (cxx_feature_check)


-- Performing Test HAVE_STD_REGEX -- success
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
```
